internal/pkg/versioncheck/detect.go

```diff
@@ -0,0 +1,226 @@
+package versioncheck
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/go-version"
+	"github.com/hashicorp/hcp/internal/pkg/api/releasesapi/client"
+	"github.com/hashicorp/hcp/internal/pkg/api/releasesapi/client/operations"
+	"github.com/hashicorp/hcp/internal/pkg/api/releasesapi/models"
+	"github.com/hashicorp/hcp/internal/pkg/iostreams"
+	hcpversion "github.com/hashicorp/hcp/version"
+	"github.com/mitchellh/go-homedir"
+)
+
+const (
+	// hcpProduct is the product name for HCP CLI in releases.
+	hcpProduct = "hcp"
+)
+
+// Checker is a version checker for the HCP CLI.
+type Checker struct {
+	io             iostreams.IOStreams
+	client         operations.ClientService
+	currentVersion *version.Version
+	checkState     *versionCheckState
+	statePath      string
+	sync.Mutex
+}
+
+// New returns a new version checker for the HCP CLI. It will used the passed
+// iostreams to display any new version information and will store the
+// stateFilePath.
+func New(io iostreams.IOStreams, stateFilePath string) (*Checker, error) {
+	return newChecker(io, stateFilePath,
+		hcpversion.FullVersion(), operations.New(client.Default.Transport, nil))
+}
+
+func newChecker(io iostreams.IOStreams, stateFilePath string,
+	currentVersion string, client operations.ClientService) (*Checker, error) {
+
+	// Parse the current version.
+	current, err := version.NewVersion(currentVersion)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse current version (%q): %w", hcpversion.FullVersion(), err)
+	}
+
+	path, err := homedir.Expand(stateFilePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed expanding HCP config directory path %q: %w", stateFilePath, err)
+	}
+
+	// Ensure the folder exists.
+	dir := filepath.Dir(path)
+	if _, err := os.Stat(dir); err != nil {
+		// If the directory doesn't exist, create it.
+		if errors.Is(err, fs.ErrNotExist) {
+			if err := os.MkdirAll(dir, 0766); err != nil {
+				return nil, fmt.Errorf("failed to created HCP config directory %q: %w", dir, err)
+			}
+		} else {
+			return nil, fmt.Errorf("failed to check if HCP config directory exists: %w", err)
+		}
+	}
```

Comment:
This will create the folders for the state file even if the version will not get checked and the state file hence not used. Should we move this into the `Check` method?

Reply:
I could go either way. I put it here because the New() is setting up the checker to successfully save its state. Gives an earlier place to handle an invalid state path.